### PR TITLE
kchat: init at 2.4.0

### PR DESCRIPTION
**Pull request by Vinetos**
## Description of changes
This MR adds the KChat from Infomaniak.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Vinetos** [comment]
Result of `nixpkgs-review pr 289983` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>kchat</li>
  </ul>
</details>

File: pkgs/by-name/kc/kchat/package.nix [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+}:
+
+appimageTools.wrapType2 rec {
+  pname = "kchat";
+  version = "2.4.0";
+
+  src = fetchurl {
+    url = "https://download.storage5.infomaniak.com/${pname}/${pname}-desktop-${version}-linux-x86_64.AppImage";
+    name = "${pname}-${version}.AppImage";
+    sha256 = "sha256-8mkkHod7iBhHVAL/vQCVnmwVlPGikdHhtiEaFVIayrU=";
```
Comment (eclairevoyant) [inline pkgs/by-name/kc/kchat/package.nix:13]:
```suggestion
    hash = "sha256-8mkkHod7iBhHVAL/vQCVnmwVlPGikdHhtiEaFVIayrU=";
```

File: pkgs/by-name/kc/kchat/package.nix [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+      mkdir -p "$out/share/lib/kchat"
+      cp -r ${contents}/{locales,resources} "$out/share/lib/kchat"
+      cp -r ${contents}/usr/* "$out"
+      cp "${contents}/${pname}-desktop.desktop" "$out/share/applications/"
+      mv "$out/bin/${pname}-${version}" "$out/bin/${pname}"
+      substituteInPlace $out/share/applications/${pname}-desktop.desktop --replace 'Exec=AppRun' 'Exec=${pname}'
+    '';
+
+  meta = with lib; {
+    description = "Instant messaging service part of Infomaniak KSuite.";
```
Comment (eclairevoyant) [inline pkgs/by-name/kc/kchat/package.nix:31]:
```suggestion
    description = "Instant messaging service part of Infomaniak KSuite";
```

File: pkgs/by-name/kc/kchat/package.nix [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    homepage = "https://www.infomaniak.com/en/apps/download-kchat";
+    license = licenses.unfree;
+    maintainers = [ maintainers.vinetos ];
+    mainProgram = "kchat";
+
```
Comment (eclairevoyant) [inline pkgs/by-name/kc/kchat/package.nix:36]:
```suggestion
```

File: pkgs/by-name/kc/kchat/package.nix [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    platforms = [ "x86_64-linux" ];
+    sourceProvenance = [ lib.sourceTypes.binaryNativeCode ];
+
```
Comment (eclairevoyant) [inline pkgs/by-name/kc/kchat/package.nix:39]:
```suggestion
```

**Review by eclairevoyant — CHANGES_REQUESTED** [review]
Thanks for the contribution! Some feedback below, and the commit/PR name should have the "v" removed

**Comment by Vinetos** [comment]
> Thanks for the contribution! Some feedback below, and the commit/PR name should have the "v" removed

Thanks. Review applied

**Comment by Vinetos** [comment]
Result of `nixpkgs-review pr 289983` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>kchat</li>
  </ul>
</details>

File: pkgs/by-name/kc/kchat/package.nix [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+, appimageTools
+}:
+
+appimageTools.wrapType2 rec {
+  pname = "kchat";
+  version = "2.4.0";
+
+  src = fetchurl {
+    url = "https://download.storage5.infomaniak.com/${pname}/${pname}-desktop-${version}-linux-x86_64.AppImage";
+    name = "${pname}-${version}.AppImage";
```
Comment (drupol) [inline pkgs/by-name/kc/kchat/package.nix:12]: Please don't use this.

`repo = pname;` is nice for DRY but creates a binding that goes too far.

See further information about this here: https://github.com/nix-community/nixpkgs-lint/issues/21

**Review by drupol — CHANGES_REQUESTED** [review]

File: pkgs/by-name/kc/kchat/package.nix (hunk 5 above)
Comment (Vinetos) [reply pkgs/by-name/kc/kchat/package.nix:12]: Hello, 
Thanks for letting me know. I was not aware of this. I am updating this.

Comment (drupol) [reply pkgs/by-name/kc/kchat/package.nix]: Don't rely on `pname` in the rest of the file, indeed.

File: pkgs/by-name/kc/kchat/package.nix [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+      contents = appimageTools.extractType2 { inherit version src; };
+    in
+    ''
+      mkdir -p "$out/share/applications"
+      mkdir -p "$out/share/lib/kchat"
+      cp -r ${contents}/{locales,resources} "$out/share/lib/kchat"
+      cp -r ${contents}/usr/* "$out"
+      cp "${contents}/kchat-desktop.desktop" "$out/share/applications/"
+      mv "$out/bin/kchat-${version}" "$out/bin/kchat"
+      substituteInPlace $out/share/applications/kchat-desktop.desktop --replace 'Exec=AppRun' 'Exec=kchat'
```
Comment (drupol) [inline pkgs/by-name/kc/kchat/package.nix:27]:
```suggestion
      mv "$out/bin/kchat-${version}" "$out/bin/${meta.mainProgram}"
      substituteInPlace $out/share/applications/kchat-desktop.desktop --replace 'Exec=AppRun' 'Exec=${meta.mainProgram}'
```

**Review by drupol — APPROVED** [review]

**Comment by Vinetos** [comment]
Result of `nixpkgs-review pr 289983` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>kchat</li>
  </ul>
</details>

**Pull request merged**